php-checks/src/test/resources/checks/UnusedPrivateMethodCheckWithTraits.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+class MagicMethodCall
+{
+  use MyTraitWithMagicMethodCallWithCallUserFuncArrayFunction;
+
+  private function bar()
+  {
+  // Compliant as bar() can be called in the trait
+  }
+}
+
+trait MyTraitWithMagicMethodCallWithCallUserFuncArrayFunction
+{
+  public function __call($method, $arguments)
+  {
+    if (method_exists($this, $method)) {
+      return call_user_func_array([$this, $method], $arguments);
+    }
+    trigger_error('Call to undefined method '.__CLASS__.'::'.$method.'()', E_USER_ERROR);
+  }
+}
+
+class MagicMethodCallFalseNegative
+{
+  use MyTraitWithMagicMethodCallWithoutBody;
+
+  // TODO SONARPHP-1481 FN because in trait in __call the call of call_user_func_array or call_user_func is missing
+  private function bar()
+  {
+  }
+}
+
+trait MyTraitWithMagicMethodCallWithoutBody
+{
+  public function __call($method, $arguments)
+  {
+  // The call of call_user_func_array or call_user_func is missing
+  }
+}
+
+class MagicMethodCall2
+{
+  use MyTraitWithoutMagicMethodCall;
+
+  private function bar() // Noncompliant
+  {
+  }
+}
+
+trait MyTraitWithoutMagicMethodCall
+{
+  public function foo()
+  {
+  }
+}
+
+class MagicMethodCall2
+{
+  use NonExistingTrait;
+
+  private function bar() // Noncompliant
+  {
+  }
+}
```

Comment:
Please add an example where trait is used in superclass.

Reply:
Good catch! Unfortunately this is not possible to detect for us, as we can't access traits in the superclass symbol (missing the AST nodes). So for this we have to live with FP's.
I will create a new ticket for this though and group the other tickets related to this rule as it's all about the same problem.